### PR TITLE
8292955: Collections.checkedMap Map.merge does not properly check key and value

### PR DESCRIPTION
**Pull request by Korov**
When the specified key did not associated with a value, should check the `key` and `value` type.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8292955](https://bugs.openjdk.org/browse/JDK-8292955): Collections.checkedMap Map.merge does not properly check key and value (**Bug** - P4)

### Reviewers
 * [Guoxiong Li](https://openjdk.org/census#gli) (@lgxbslgx - **Reviewer**)
 * [Chen Liang](https://openjdk.org/census#liach) (@liach - Author)
 * [Per Minborg](https://openjdk.org/census#pminborg) (@minborg - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18141/head:pull/18141` \
`$ git checkout pull/18141`

Update a local copy of the PR: \
`$ git checkout pull/18141` \
`$ git pull https://git.openjdk.org/jdk.git pull/18141/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18141`

View PR using the GUI difftool: \
`$ git pr show -t 18141`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18141.diff">https://git.openjdk.org/jdk/pull/18141.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18141#issuecomment-1981537332)

### CODE REVIEW

File: test/jdk/java/util/Collections/CheckedMapBash.java
Comment (lgxbslgx) [inline test/jdk/java/util/Collections/CheckedMapBash.java:192]: May be better to seperate these two cases into two tests. Because if the first case fails, the latter can't be verified now.

**Review by lgxbslgx — CHANGES_REQUESTED** [review]
Good caught. A trivial suggestion.

**Comment by Korov** [comment]
> Good caught. A trivial suggestion.

Thanks for your suggestion, the code has been modified.

File: test/jdk/java/util/Collections/CheckedMapBash.java
Comment (liach) [inline test/jdk/java/util/Collections/CheckedMapBash.java:193]: Since this test is already running with testNG, we should just use its builtin functionalities:
```java
    @Test(groups = "type_check")
    public static void testTypeCheck() {
        Map m = Collections.checkedMap(new HashMap<>(), Integer.class, Integer.class);
        Assert.assertThrows(ClassCastException.class, () -> m.merge("key", "value", (v1, v2) -> null));
        Assert.assertThrows(ClassCastException.class, () -> m.merge("key", 3, (v1, v2) -> v2));
    }
```

In addition, the checked map tests are in general, only checking the map functionality works without checking if the type checking functionality works. But adding a comprehensive test for checking functionalites would be much more work and is out of this patch's scope, so you don't need to worry about it.

Comment (Korov) [reply test/jdk/java/util/Collections/CheckedMapBash.java]: Thank you very much for informing me about these things. I have modified the test function name.

**Review by lgxbslgx — APPROVED** [review]
Looks good.

**Comment by Korov** [comment]
Hi @liach , Can you review my newly submitted code?

**Review by liach — APPROVED** [review]
Looks good to me, maybe @stuart-marks can take a look too?

**Comment by liach** [comment]
Keep-alive, maybe people like @viktorklang-ora might look at this simple fix too

**Comment by viktorklang-ora** [comment]
@liach I'm not a Reviewer (yet) so I'll defer to someone like @stuart-marks :)

**Comment by Korov** [comment]
Hi @stuart-marks , Can you take a moment to review my PR? thank you very much for your help!

**Comment by liach** [comment]
@viktorklang-ora Can you just leave an approval as a collections engineer to ensure this patch isn't "accidetnally integrated"?

/reviewers 2 committer

**Comment by liach** [comment]
@minborg Is it possible for you to review this collection bugfix?

**Comment by minborg** [comment]
> @minborg Is it possible for you to review this collection bugfix?

Will do! Thanks for the heads up.

**Review by minborg — APPROVED** [review]
Adding the checks before remapping seems the right thing to do in order to uphold the checked maps guarantees on types and values.

**Pull request closed without merging**
